src/pip/_internal/resolution/resolvelib/requirements.py

```diff
@@ -1,141 +1,110 @@
 from pip._vendor.packaging.utils import canonicalize_name
 
-from pip._internal.req.req_install import InstallRequirement
 from pip._internal.utils.typing import MYPY_CHECK_RUNNING
 
-from .base import Candidate, Requirement, format_name
+from .base import Requirement
+from .candidates import make_candidate
 
 if MYPY_CHECK_RUNNING:
-    from typing import (Optional, Sequence)
-
-    from pip._vendor.packaging.version import _BaseVersion
+    from typing import Sequence
 
     from pip._internal.index.package_finder import PackageFinder
-
-
-def make_requirement(install_req):
-    # type: (InstallRequirement) -> Requirement
-    if install_req.link:
-        if install_req.req and install_req.req.name:
-            return NamedRequirement(install_req)
-        else:
-            return UnnamedRequirement(install_req)
+    from pip._internal.operations.prepare import RequirementPreparer
+    from pip._internal.req.req_install import InstallRequirement
+    from pip._internal.resolution.base import InstallRequirementProvider
+
+    from .base import Candidate
+
+
+def make_requirement(
+    ireq,      # type: InstallRequirement
+    finder,    # type: PackageFinder
+    preparer,  # type: RequirementPreparer
+    make_install_req  # type: InstallRequirementProvider
+):
+    # type: (...) -> Requirement
+    if ireq.link:
+        candidate = make_candidate(
+            ireq.link,
+            preparer,
+            ireq,
+            make_install_req
+        )
+        return ExplicitRequirement(candidate)
     else:
-        return VersionedRequirement(install_req)
+        return SpecifierRequirement(
+            ireq,
+            finder,
+            preparer,
+            make_install_req
+        )
 
 
-class UnnamedRequirement(Requirement):
-    def __init__(self, req):
-        # type: (InstallRequirement) -> None
-        self._ireq = req
-        self._candidate = None  # type: Optional[Candidate]
+class ExplicitRequirement(Requirement):
+    def __init__(self, candidate):
+        # type: (Candidate) -> None
+        self.candidate = candidate
 
     @property
     def name(self):
         # type: () -> str
-        assert self._ireq.req is None or self._ireq.name is None, \
-            "Unnamed requirement has a name"
-        # TODO: Get the candidate and use its name...
-        return ""
-
-    def _get_candidate(self):
-        # type: () -> Candidate
-        if self._candidate is None:
-            self._candidate = Candidate()
-        return self._candidate
-
-    def find_matches(
-        self,
-        finder,     # type: PackageFinder
-    ):
-        # type: (...) -> Sequence[Candidate]
-        return [self._get_candidate()]
+        return self.candidate.name
```

Comment:
I assume this implies candidate name is always canonical. Maybe add a note somewhere to make it clear?

Reply:
Done